### PR TITLE
Unexpected Browser Openings - Fix to Issue-1044

### LINKED ISSUE #1044: Reporters open when not supposed to (and nameless file tested)
I have just discovered, running `pytest --cov python_ta --cov-config=.coveragerc tests`, which produces the failure mentioned in Issue #1040, does NOT open any `python_ta` reporter windows (as desired). 

However, running `pytest --cov python_ta --cov-config=.coveragerc tests --ignore=tests/test_check.py` DOES result in no less than 4 `python_ta` reporter windows opening in browser automatically. 

When running without coverage, we get the exact same results.

`pytest tests` produces no `python_ta` reporters, but `pytest tests --ignore=tests/test_check.py` produces the same 4 reporters as when we ran with coverage.

These 4 reporter windows are as follows:
- _a file with no name_, `pyta/tests/test_config/file_fixtures/test_with_errors.pylintrc`, and `examples/nodes/name.py`
- `pyta/tests/test_config/file_fixtures/test_f0011.pylintrc` and `examples/nodes/name.py` (yes, both files in the same reporter window; this window opens twice)
- `pyta/tests/test_custom_checkers/file_fixtures/badModuleName.py`

<details>
<summary>Visual aids</summary>

Four reports opened:
four Python TA reports opened as tabs in browser

Nameless file:
nameless file in table of contents
nameless file in main reporter feed
</details>

The fact that `pytest` is somehow getting called on a file that has no name seems like it would be worth looking into.

### PR DESCRIPTION
**Pull request by CulmoneY**
# Proposed Changes

As described in Issue #1044, when running `pytest tests --ignore=tests/test_check.py`, 4 unexpected pythonTA web browser reports open. However, whilst running only `pytest tests`, no reporters open as expected. This PR fixes an issue where the python_ta.check_all() function was opening a web browser and handling HTTP requests during test execution, which is not necessary for unit tests.

To resolve this issue, I introduced mocking for webbrowser.open and HTTPServer.handle_request to prevent unwanted browser tabs and HTTP request handling during test runs. This mocking is applied in tests that call python_ta.check_all() [which did not report data as a json file], ensuring that reports are generated without opening the browser or interacting with the server.

I created a new Pytest fixture, `prevent_webbrowser_and_httpserver `, to handle the mocking. This fixture was selectively applied to tests that required it. fix #1044 
</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |     X     |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: tests/test_custom_checkers/test_invalid_name_checker.py [hunk 1]
```diff
@@ -639,7 +639,7 @@ def snake_case_format():
             self.checker.visit_functiondef(functiondef_node)
 
 
-def test_module_name_no_snippet() -> None:
+def test_module_name_no_snippet(prevent_webbrowser_and_httpserver) -> None:
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_invalid_name_checker.py:642]: This test is a bit different than the others. It should be possible to modify the test to pass in the default test pylintrc file, which I believe uses a different reporter than the HTML reporter.

File: tests/conftest.py [hunk 2]
```diff
@@ -9,3 +13,11 @@ def disable_contract_checking():
     python_ta.contracts.ENABLE_CONTRACT_CHECKING = False
     yield
     python_ta.contracts.ENABLE_CONTRACT_CHECKING = True
+
+
+@pytest.fixture()
+def prevent_webbrowser_and_httpserver():
```
Comment (david-yz-liu) [inline tests/conftest.py:19]: It's great that you've identified the relevant mocking code. Pulling it out into a fixture is a good idea, but this is revealing another problem with this code: it doesn't "reset" the functions to the original behaviour after the test case ends. This defeats the purpose of mocking, which should be tied to specific test cases without changing any global behaviour.

You can take this opportunity to add `pytest-mock` (https://github.com/pytest-dev/pytest-mock/) as a development dependency for the project, and then use it to set up this mocking. That library should correctly handle restoring the original behaviour after the tests end.

**Review by david-yz-liu — COMMENTED** [review]
@CulmoneY good work. In addition to the inline comments I left, you should review the note in the PR template about linking pull requests to issues and update your PR accordingly. More generally, you shouldn't delete the instructions found in the PR template, and instead just fill in the empty sections underneath these instructions.

**Comment by CulmoneY** [comment]
Hi David, Unfortunately, I was unable to implement `pytest-mock` as you recommended. However, I was able to resolve the issue by manually resetting `webbrowser.open` and `HTTPServer.handle_request` back to their original values after the tests. This approach seems to work as intended, and I believe it serves as a suitable alternative.

**Comment by david-yz-liu** [comment]
@CulmoneY push what you tried to do with `pytest-mock` and I'll take a look.

**Comment by CulmoneY** [comment]
> @CulmoneY push what you tried to do with `pytest-mock` and I'll take a look.

I redid the code with pytest-mock, and it worked as expected. It seems I must have had an unnoticed mistake somewhere else. Sorry!

File: pyproject.toml [hunk 3]
```diff
@@ -36,6 +36,7 @@ dev = [
     "pytest-cov >= 4.0,< 5.1",
     "sphinx",
     "sphinx-rtd-theme",
+    "pytest-mock",
```
Comment (david-yz-liu) [inline pyproject.toml:39]: Move this in alphabetical order in this list

File: tests/conftest.py [hunk 4]
```diff
@@ -1,3 +1,7 @@
+import webbrowser
+from http.server import HTTPServer
+from unittest.mock import Mock
```
Comment (david-yz-liu) [inline tests/conftest.py:3]: I think this is no longer used

File: tests/conftest.py [hunk 5]
```diff
@@ -9,3 +13,11 @@ def disable_contract_checking():
     python_ta.contracts.ENABLE_CONTRACT_CHECKING = False
     yield
     python_ta.contracts.ENABLE_CONTRACT_CHECKING = True
+
+
+@pytest.fixture()
+def prevent_webbrowser_and_httpserver(mocker):
+    """Automatically mock webbrowser.open and HTTPServer.handle_request in all tests. this prevents any owser/server
```
Comment (david-yz-liu) [inline tests/conftest.py:20]: there's a typo in the word "owser"

File: tests/test_custom_checkers/test_invalid_name_checker.py [hunk 6]
```diff
@@ -643,8 +643,8 @@ def test_module_name_no_snippet() -> None:
     """Test that PythonTA does not build a snippet for the message added by this checker."""
     curr_dir = os.path.dirname(__file__)
     file_fixture = os.path.join(curr_dir, "file_fixtures", "badModuleName.py")
-    reporter = python_ta.check_all(module_name=file_fixture)
+    config = os.path.join(os.path.dirname(curr_dir), "test.pylintrc")
+    reporter = python_ta.check_all(module_name=file_fixture, config=config)
 
     snippet = reporter.messages[file_fixture][0].snippet
-
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_invalid_name_checker.py:649]: revert this change

File: CHANGELOG.md [hunk 7]
```diff
@@ -24,6 +24,7 @@ and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### 🔧 Internal changes
 
+- Added mock `webbrowser.open` in tests to prevent browser tabs and HTTP requests during `python_ta.check_all()` executions.
```
Comment (david-yz-liu) [inline CHANGELOG.md:27]: please add this next to the other entry

File: tests/conftest.py (hunk 4 above)
Comment (CulmoneY) [reply tests/conftest.py:3]: Everything except `from http.server import HTTPServer` has been removed.

**Review by david-yz-liu — APPROVED** [review]
